subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -1199,203 +1186,75 @@ public void stop() {
         }
     }
 
-    interface TypeSpec extends Spec<Type> {
-        Type getType();
-    }
-
-    private static class ClassSpec implements TypeSpec {
-        private final Class<?> type;
-
-        private ClassSpec(Class<?> type) {
-            this.type = type;
-        }
-
-        public Type getType() {
-            return type;
-        }
-
-        public boolean isSatisfiedBy(Type element) {
-            if (element instanceof ParameterizedType) {
-                ParameterizedType parameterizedType = (ParameterizedType) element;
-                if (parameterizedType.getRawType() instanceof Class) {
-                    return type.isAssignableFrom((Class) parameterizedType.getRawType());
+    public ServiceProvider find(Type serviceType, Provider provider) {
+        if (serviceType instanceof ParameterizedType) {
```

Comment:
It would help readability if all those branches were turned into their own methods.